### PR TITLE
[BEAM-2975] Clarify semantics of objects returned by state access

### PR DESCRIPTION
**Pull request by dpmills**
R: @tgroh

### CODE REVIEW

**Comment by dpmills** [comment]
run flink ValidatesRunner

**Comment by dpmills** [comment]
run spark ValidatesRunner

File: sdks/java/core/src/main/java/org/apache/beam/sdk/state/ReadableState.java [hunk 1]
```diff
@@ -36,6 +36,8 @@
    * <p>If there will be many calls to {@link #read} for different state in short succession, you
    * should first call {@link #readLater} for all of them so the reads can potentially be batched
    * (depending on the underlying implementation}.
+   *
+   * <p>Subclasses should not modify the returned object through other calls to the state object.
```
Comment (tgroh) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/state/ReadableState.java]: Perhaps "The returned object should be independent of the underlying state."? Any `T` returned should not perform write-backs without going through the appropriate state interface, and modification to the state should not be mirrored in the returned object.

Comment (dpmills) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/state/ReadableState.java]: Done

**Pull request closed without merging**